From 334b441cc811e66f834780fd696917439ec1137d Mon Sep 17 00:00:00 2001
From: Nick Xie <nick@khadas.com>
Date: Mon, 2 Feb 2026 11:00:40 +0800
Subject: [PATCH 180/210] arm64: dts: amlogic: meson-s4: add USB2 support

Add USB2 support to S4 SoC family

Signed-off-by: Nick Xie <nick@khadas.com>
[chewitt: add the USB controller node from the DRM support commit, and
 enable USB host on the Khadas VIM1S with a regulator switching VBUS on]
Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
 .../amlogic/meson-s4-s905y4-khadas-vim1s.dts  | 23 ++++++++++
 arch/arm64/boot/dts/amlogic/meson-s4.dtsi     | 46 +++++++++++++++++++
 2 files changed, 69 insertions(+)

diff --git a/arch/arm64/boot/dts/amlogic/meson-s4-s905y4-khadas-vim1s.dts b/arch/arm64/boot/dts/amlogic/meson-s4-s905y4-khadas-vim1s.dts
index fd265764f499..9595fa274614 100644
--- a/arch/arm64/boot/dts/amlogic/meson-s4-s905y4-khadas-vim1s.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-s4-s905y4-khadas-vim1s.dts
@@ -135,6 +135,16 @@ vddao_3v3: regulator-vddao-3v3 {
 		regulator-always-on;
 	};
 
+	usb_pwr: regulator-usb-pwr {
+		compatible = "regulator-fixed";
+		regulator-name = "USB_PWR";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		vin-supply = <&main_5v>;
+		gpio = <&gpio GPIOH_6 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
 	vddio_ao1v8: regulator-vddio-ao1v8 {
 		compatible = "regulator-fixed";
 		regulator-name = "VDDIO_AO1V8";
@@ -390,3 +400,16 @@ bluetooth {
 &uart_b {
 	status = "okay";
 };
+
+&usb {
+	status = "okay";
+};
+
+&usb2_phy0 {
+	status = "okay";
+};
+
+&usb2_phy1 {
+	status = "okay";
+	phy-supply = <&usb_pwr>;
+};
diff --git a/arch/arm64/boot/dts/amlogic/meson-s4.dtsi b/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
index a7c2fc91044e..8107b7f1f9ff 100644
--- a/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
+++ b/arch/arm64/boot/dts/amlogic/meson-s4.dtsi
@@ -872,6 +872,28 @@ uart_b: serial@7a000 {
 				status = "disabled";
 			};
 
+			usb2_phy0: phy@3c000 {
+				compatible = "amlogic,s4-usb2-phy";
+				reg = <0x0 0x3c000 0x0 0x2000>;
+				clocks = <&xtal>;
+				clock-names = "xtal";
+				resets = <&reset RESET_USBPHY20>;
+				reset-names = "phy";
+				#phy-cells = <0>;
+				status = "disabled";
+			};
+
+			usb2_phy1: phy@3e000 {
+				compatible = "amlogic,s4-usb2-phy";
+				reg = <0x0 0x3e000 0x0 0x2000>;
+				clocks = <&xtal>;
+				clock-names = "xtal";
+				resets = <&reset RESET_USBPHY21>;
+				reset-names = "phy";
+				#phy-cells = <0>;
+				status = "disabled";
+			};
+
 			reset: reset-controller@2000 {
 				compatible = "amlogic,meson-s4-reset";
 				reg = <0x0 0x2000 0x0 0x98>;
@@ -983,6 +1005,30 @@ emmc: mmc@fe08c000 {
 			assigned-clock-rates = <24000000>;
 		};
 
+		usb: usb@fe03a000 {
+			compatible = "amlogic,meson-g12a-usb-ctrl";
+			reg = <0x0 0xfe03a000 0x0 0xa0>;
+			interrupts = <GIC_SPI 129 IRQ_TYPE_LEVEL_HIGH>;
+			#address-cells = <2>;
+			#size-cells = <2>;
+			ranges;
+			clocks = <&clkc_periphs CLKID_USB>;
+			resets = <&reset RESET_USBCTRL>;
+			power-domains = <&pwrc PWRC_S4_USB_COMB_ID>;
+			dr_mode = "host";
+			phys = <&usb2_phy0>, <&usb2_phy1>;
+			phy-names = "usb2-phy0", "usb2-phy1";
+			status = "disabled";
+
+			crg_xhci: usb@fde00000 {
+				compatible = "generic-xhci";
+				status = "okay";
+				reg = <0x0 0xfde00000 0x0 0x100000>;
+				interrupts = <GIC_SPI 130 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&clkc_periphs CLKID_USB1_TO_DDR>;
+			};
+		};
+
 		di: video-deinterlacer@ff005c00 {
 			compatible = "amlogic,s4-di";
 			reg = <0x0 0xff005c00 0x0 0x10000>;
-- 
2.43.0

